scylla/src/cloud/config.rs

```diff
@@ -179,4 +212,262 @@ mod deserialise {
+                username: auth_info.username,
+                password: auth_info.password,
+            })
+        }
+    }
+
+    impl TryFrom<Datacenter> for super::Datacenter {
+        type Error = CloudConfigError;
+
+        fn try_from(datacenter: Datacenter) -> Result<Self, Self::Error> {
+            // Validate node domain
+            // Using parts relevant to hostnames as we're dealing with a part of hostname
+            // RFC-1123 Section 2.1 and RFC-952 1.
+            let node_domain = datacenter.nodeDomain;
+            if node_domain.is_empty() {
+                return Err(CloudConfigError::Validation(
+                    "nodeDomain property is required in datacenter description.".into(),
+                ));
+            }
+            if node_domain.len() > NODE_DOMAIN_MAX_LENGTH {
+                return Err(CloudConfigError::Validation(format!(
+                    "Subdomain name too long (max {} ): {}",
+                    NODE_DOMAIN_MAX_LENGTH, &node_domain
+                )));
+            }
+            if node_domain.contains(' ') {
+                return Err(CloudConfigError::Validation(format!(
+                    "nodeDomain {} cannot contains spaces.",
+                    &node_domain
+                )));
+            }
+            if node_domain.starts_with('.') || node_domain.ends_with('.') {
+                return Err(CloudConfigError::Validation(format!(
+                    "nodeDomain {} cannot start or end with a dot.",
+                    &node_domain
+                )));
+            }
+            if node_domain.ends_with('-') {
+                return Err(CloudConfigError::Validation(format!(
+                    "nodeDomain {} cannot end with a minus sign.",
+                    &node_domain
+                )));
+            }
+
+            let components = node_domain.split('.');
+            for component in components {
+                if component.is_empty() {
+                    return Err(CloudConfigError::Validation(format!(
+                        "nodeDomain {} cannot have empty components between dots.",
+                        &node_domain
+                    )));
+                }
+
+                if component.starts_with('-') || component.ends_with('-') {
+                    return Err(CloudConfigError::Validation(format!(
+                        "nodeDomain {} components can have minus sign only as interior character,\
+                            which component {} disobeys.",
+                        &node_domain, component,
+                    )));
+                }
+
+                for c in component.chars() {
+                    if !c.is_alphanumeric() && c != '-' {
+                        return Err(CloudConfigError::Validation(format!(
+                            "nodeDomain {} contains illegal character: {}.",
+                            &node_domain, c
+                        )));
+                    }
+                }
+            }
```

Comment:
I've followed Java driver's strict validation. As they have done this meticulously, I've decided to use their logic instead of relying on completeness of a library's assertions.

Reply:
I think it is hard to beat an external library in this regard. `url` is a part of the servo project, looks maintained and exists for some time (9 years) with 83M downloads, so it is most likely battle tested.

On the other hand, the RFCs mentioned here are quite old and if we don't need to support anything newer than them, then it may be fine. To be fair, the exact requirements/restrictions for this field are not 100% clear to me.

Reply:
I'll stick to the already written manual validation.